flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/reader/BulkFormat.java

```diff
@@ -0,0 +1,158 @@
+ * {@link BulkFormat#createReader(Configuration, Path, long, long)} or
+ * {@link BulkFormat#createReader(Configuration, Path, long, long, long)} methods.
+ * The outer class acts mainly as a configuration holder and factory for the reader.
+ *
+ * <h2>Checkpointing</h2>
+ *
+ * <p>The bulk reader returns an iterator structure per batch. The iterator produces records together
+ * with a position. That position is the point from where the reading can be resumed AFTER
+ * the records was emitted. So that position points effectively to the record AFTER the current record.
+ *
+ * <p>The simplest way to return this position information is to always assume a zero offset in the file
+ * and simply increment the record count. Note that in this case the fist record would be returned with
+ * a record count of one, the second one with a record count of two, etc.
+ *
+ * <p>Formats that have the ability to efficiently seek to a record (or to every n-th record) by offset
+ * in the file can work with the position field to avoid having to read and discard many records on recovery.
+ *
+ * <h2>Serializable</h2>
+ *
+ * <p>Like many other API classes in Flink, the outer class is serializable to support sending instances
+ * to distributed workers for parallel execution. This is purely short-term serialization for RPC and
+ * no instance of this will be long-term persisted in a serialized form.
+ *
+ * <h2>Record Batching</h2>
+ *
+ * <p>Internally in the file source, the readers pass batches of records from the reading
+ * threads (that perform the typically blocking I/O operations) to the async mailbox threads that
+ * do the streaming and batch data processing. Passing records in batches (rather than one-at-a-time)
+ * much reduce the thread-to-thread handover overhead.
+ *
+ * <p>For the {@code BulkFormat}, one batch (as returned by {@link BulkFormat.Reader#readBatch()}) is
+ * handed over as one.
+ */
+public interface BulkFormat<T> extends Serializable {
+
+	/**
+	 * Creates a new reader that reads from {@code filePath} starting at {@code offset} and reads
+	 * until {@code length} bytes after the offset.
+	 */
+	default BulkFormat.Reader<T> createReader(Configuration config, Path filePath, long offset, long length) throws IOException {
+		return createReader(config, filePath, offset, length, 0L);
+	}
+
+	/**
+	 * Creates a new reader that reads from {@code filePath} starting at {@code offset} and reads
+	 * until {@code length} bytes after the offset. A number of {@code recordsToSkip} records should be
+	 * read and discarded after the offset. This is typically part of restoring a reader to a checkpointed
+	 * position.
+	 */
+	BulkFormat.Reader<T> createReader(Configuration config, Path filePath, long offset, long length, long recordsToSkip) throws IOException;
+
+	/**
+	 * Gets the type produced by this format. This type will be the type produced by the file
+	 * source as a whole.
+	 */
+	TypeInformation<T> getProducedType();
```

Comment:
I am torn on this one. The `ResultTypeQueryable` is mainly used by the `StreamExecutionEnviroment` and `DataStream` classes. The formats here are not used directly with those.

Reply:
Personally I feel it is OK to use `ResultTypeQueryable` here. The class is in `org.apache.flink.api.java.typeutils` package which seems up for general usage.

Reply:
I can add it, if you feel strongly about this. I am usually a bit cautious adding more interfaces, because especially in a public API one can never undo this again.

Reply:
I think you are right. It makes sense to add the interface. I guess my main reservation was that the name `ResultTypeQueryable` is not a great name. But that shouldn't stop anything here.

Reply:
I fine either way. I just feel that it would be preferable to follow the same pattern when doing the same thing. So the code will be in a consistent style across modules.